### PR TITLE
Added nimble arduino file

### PR DESCRIPTION
**Pull request by JackChin3**

### CODE REVIEW

File: _arduino/nimbleProportionalControl.ino [hunk 1]
```diff
@@ -0,0 +1,391 @@
+#include <NimBLEDevice.h>
+#include <Chrono.h>        // http://github.com/SofaPirate/Chrono
+#include <ESP32Encoder.h>  // https://github.com/madhephaestus/ESP32Encoder.git
+
+//  for OLED and chirps 
+#include <Arduino.h>
+#include <U8x8lib.h> // need to install u8g2 by oliver to access this library 
+#include <Wire.h>
```
Comment (anthonyjclark) [inline _arduino/nimbleProportionalControl.ino]: Add a README.md to this directory and add these in a list of dependencies. Then remove the comments from here.

File: _arduino/nimbleProportionalControl.ino [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+
+#define WHEEL_CIRCUMFERENCE_MM 251.32
+#define PULSES_PER_ROTATION 360 // TODO: 7*2*50?
+#define MAX_SPEED_MM_PER_S 800
+
+#define K_P .05
+#define PWM_MAX_DELTA_PER_UPDATE 20
+#define CONTROL_UPDATE_PERIOD_MS 50
+#define CONTROL_UPDATE_PERIOD_S (float(CONTROL_UPDATE_PERIOD_MS) / 1000.0)
+#define SPEED_THRESH 5
```
Comment (anthonyjclark) [inline _arduino/nimbleProportionalControl.ino]: What are the units?

File: _arduino/nimbleProportionalControl.ino [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+
+U8X8_SSD1306_128X64_NONAME_HW_I2C u8x8(SCL, SDA, U8X8_PIN_NONE);   // OLEDs without Reset of the Display
```
Comment (anthonyjclark) [inline _arduino/nimbleProportionalControl.ino]:
```suggestion
// OLEDs without Reset of the Display
U8X8_SSD1306_128X64_NONAME_HW_I2C u8x8(SCL, SDA, U8X8_PIN_NONE); 
```

File: _arduino/nimbleProportionalControl.ino [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+int speakerPin = D6;
```
Comment (anthonyjclark) [inline _arduino/nimbleProportionalControl.ino]: Should be a `#define`

File: _arduino/nimbleProportionalControl.ino [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+      
+    int userSetSpeed = map(motorSpeedPercent, -100, 100, -MAX_SPEED_MM_PER_S, MAX_SPEED_MM_PER_S);
+    //   compute e = userSetSpeed - measureSpeed
+    float error = userSetSpeed - measuredSpeed;
+    //   compute uDelta = constrain(kp*e, -PWM_MAX_DELTA_PER_UPDATE, PWM_MAX_DELTA_PER_UPDATE) 
+    return constrain(K_P * error, -PWM_MAX_DELTA_PER_UPDATE, PWM_MAX_DELTA_PER_UPDATE);
+
+  }
+
+};
```
Comment (anthonyjclark) [inline _arduino/nimbleProportionalControl.ino]: Format and move into its own file.

File: _arduino/nimbleProportionalControl.ino [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+// ▐█ █▌      ▐▌                 
+// ▐███▌ ▟█▙ ▐███  ▟█▙  █▟█▌▗▟██▖
+// ▐▌█▐▌▐▛ ▜▌ ▐▌  ▐▛ ▜▌ █▘  ▐▙▄▖▘
+// ▐▌▀▐▌▐▌ ▐▌ ▐▌  ▐▌ ▐▌ █    ▀▀█▖
+// ▐▌ ▐▌▝█▄█▘ ▐▙▄ ▝█▄█▘ █   ▐▄▄▟▌
+// ▝▘ ▝▘ ▝▀▘   ▀▀  ▝▀▘  ▀    ▀▀▀
+// ----------------------------------------------------------------
+
+// Motor globals
+int MIN_PWM_VALUE = 0;
```
Comment (anthonyjclark) [inline _arduino/nimbleProportionalControl.ino]: Move the other motor globals here too.

File: _arduino/nimbleProportionalControl.ino [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+// ----------------------------------------------------------------
+
+// Motor globals
+int MIN_PWM_VALUE = 0; 
+
+int motorRightPWMValue = 0; 
+int motorLeftPWMValue = 0;
+
+int motorRightSpeedPercent;
+int motorLeftSpeedPercent;
```
Comment (anthonyjclark) [inline _arduino/nimbleProportionalControl.ino]: Change the name to indicate it comes from user input?

File: _arduino/nimbleProportionalControl.ino [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+
+int uLeft;
+int uRight;
```
Comment (anthonyjclark) [inline _arduino/nimbleProportionalControl.ino]: Better name.

File: _arduino/nimbleProportionalControl.ino [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+  void onWrite(NimBLECharacteristic* pSpeedCharacteristic) {
+    float speed = (pSpeedCharacteristic->getValue())[0];  // Get the value of the speed characteristic as a flloat
+     if (speed > 155) {
+        speed = speed - 256;
+      }
+      if (isRightMotorCallback_) {
+        motorRightSpeedPercent = speed;
+      } else {
+        motorLeftSpeedPercent = speed;
+      }
```
Comment (anthonyjclark) [inline _arduino/nimbleProportionalControl.ino]: Figure this out before merging.

File: _arduino/nimbleProportionalControl.ino [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+void playTone(int tone, int duration) {
+  for (long i = 0; i < duration * 1000L; i += tone * 2) {
+    digitalWrite(speakerPin, HIGH);
+    delayMicroseconds(tone);
+    digitalWrite(speakerPin, LOW);
+    delayMicroseconds(tone);
+  }
+}
+
+void playNote(char note, int duration) {
```
Comment (anthonyjclark) [inline _arduino/nimbleProportionalControl.ino]: Does playing a note work?

File: _arduino/nimbleProportionalControl.ino [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+
+    pServer = NimBLEDevice::createServer();
+    pServer->setCallbacks(new ServerCallbacks());
+
+    NimBLEService* pService = pServer->createService(SERVICE_UUID);
+
+    pMeasuredSpeedCharacteristicRight = pService->createCharacteristic(MEASURED_SPEED_CHARACTERISTIC_RIGHT_UUID, NIMBLE_PROPERTY::READ | NIMBLE_PROPERTY::NOTIFY);
+    pMeasuredSpeedCharacteristicLeft = pService->createCharacteristic(MEASURED_SPEED_CHARACTERISTIC_LEFT_UUID, NIMBLE_PROPERTY::READ | NIMBLE_PROPERTY::NOTIFY);
+
+    // TODO: do we need notify on this side of the connection?
```
Comment (anthonyjclark) [inline _arduino/nimbleProportionalControl.ino]:
```suggestion
```

File: _arduino/nimbleProportionalControl.ino [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+    encoderLeft.setCount(0);
+
+    float rightRotations = newRightPosition / (float)PULSES_PER_ROTATION;
+    float leftRotations = newLeftPosition / (float)PULSES_PER_ROTATION;
+
+    float rightDistance = rightRotations * WHEEL_CIRCUMFERENCE_MM;
+    float leftDistance = leftRotations * WHEEL_CIRCUMFERENCE_MM;
+
+    // TODO: shouldn't need to multiply by -1 here
+    float measuredSpeedRight = (rightDistance / CONTROL_UPDATE_PERIOD_S) * -1;
```
Comment (anthonyjclark) [inline _arduino/nimbleProportionalControl.ino]: Get rid of the `-1` (might need to change wiring).

File: _arduino/nimbleProportionalControl.ino [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+    pMeasuredSpeedCharacteristicLeft->setValue<float>(measuredSpeedLeft);
+    pMeasuredSpeedCharacteristicLeft->notify();
+
+    u8x8.clear();
+    u8x8.print("left encoder speed: ");
+    u8x8.print(measuredSpeedLeft);
+    u8x8.print("\n");
+    u8x8.print("right encoder speed: ");
+    u8x8.print(measuredSpeedRight);
+    u8x8.print("\n");
```
Comment (anthonyjclark) [inline _arduino/nimbleProportionalControl.ino]: This should be done at its own refresh rate.

File: _arduino/nimbleProportionalControl.ino [hunk 14]
```diff
@@ -0,0 +1,391 @@
...
+    motorRightController.set(directionValueRight, motorRightPWMValue);
+
+  }
+
+  
+  if (!deviceConnected && oldDeviceConnected) {  // device was connected but no longer true
+    Serial.println("Device disconnected.");
+    // turns both motors off immediately if it is disconnected
+    motorRightController.set(REVERSE, 0);
+    motorLeftController.set(REVERSE, 0);
```
Comment (anthonyjclark) [inline _arduino/nimbleProportionalControl.ino]: Why `REVERSE`?

**Review by anthonyjclark — CHANGES_REQUESTED** [review]

File: _arduino/nimbleProportionalControl.ino (hunk 10 above)
Comment (JackChin3) [reply _arduino/nimbleProportionalControl.ino]: Works occasionally, but may be having PIN issues.

File: _arduino/README.md [hunk 15]
```diff
@@ -0,0 +1,12 @@
+# Arduino Scripts for Motor Control and Bluetooth
+
+This arduino script sets up a BLE server and allows users to connect to the robot and controll the motors through a website.
+
+# Dependencies:
+
+- NimBLEDevice
```
Comment (anthonyjclark) [inline _arduino/README.md]:
```suggestion
- [NimBLE-Arduino](https://github.com/h2zero/NimBLE-Arduino)
```

File: _arduino/README.md [hunk 16]
```diff
@@ -0,0 +1,12 @@
+# Arduino Scripts for Motor Control and Bluetooth
```
Comment (anthonyjclark) [inline _arduino/README.md]:
```suggestion
# Arduino Scripts for the CS 181R Robot
```

File: _arduino/README.md [hunk 17]
```diff
@@ -0,0 +1,12 @@
...
+This arduino script sets up a BLE server and allows users to connect to the robot and controll the motors through a website.
+
+# Dependencies:
+
+- NimBLEDevice
+- Chrono (http://github.com/SofaPirate/Chrono)
+- ESP32Encoder (https://github.com/madhephaestus/ESP32Encoder.git)
+- Arduino
+- U8x8lib (package: u8g2 by Oliver)
+- Wire
```
Comment (anthonyjclark) [inline _arduino/README.md]:
```suggestion
- [Chrono](http://github.com/SofaPirate/Chrono)
- [ESP32Encoder](https://github.com/madhephaestus/ESP32Encoder.git)
- [U8g2_Arduino](https://github.com/olikraus/U8g2_Arduino)
```

File: _arduino/motorControl.ino [hunk 18]
```diff
@@ -0,0 +1,50 @@
+// contains a class that can be used to control the motors with proportional control 
+// #define PWM_MAX_DELTA_PER_UPDATE 20 // pwm 
+// #define MAX_SPEED_MM_PER_S 800
```
Comment (anthonyjclark) [inline _arduino/motorControl.ino]:
```suggestion
```

File: _arduino/motorControl.ino [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+
```
Comment (anthonyjclark) [inline _arduino/motorControl.ino]: Need a header guard at the top of the file.

File: _arduino/motorControl.ino [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+    int directionValue = direction == FORWARD ? forwardLevel_ : !forwardLevel_;
+    if (pwm < 0) {
+      digitalWrite(directionPin_, !directionValue);
+      analogWrite(pwmPin_, abs(pwm));
+    }
+    else {
+      digitalWrite(directionPin_, directionValue);
+      analogWrite(pwmPin_, pwm);
+    }
+  }
```
Comment (anthonyjclark) [inline _arduino/motorControl.ino]: `pwm` should never be `< 0`

File: _arduino/nimbleProportionalControl.ino [hunk 21]
```diff
@@ -0,0 +1,342 @@
+#include <NimBLEDevice.h>
+#include <Chrono.h>
+#include <ESP32Encoder.h>
+#include <Arduino.h>
+#include <U8x8lib.h>
+#include <Wire.h>
+
+#define CONTROL_UPDATE_PERIOD_MS 50 // miliseconds 
+#define OLED_UPDATE_PERIOD_MS 500 // miliseconds
+#define CONTROL_UPDATE_PERIOD_S (float(CONTROL_UPDATE_PERIOD_MS) / 1000.0) // seconds
```
Comment (anthonyjclark) [inline _arduino/nimbleProportionalControl.ino]: Swap these two lines.

File: _arduino/nimbleProportionalControl.ino [hunk 22]
```diff
@@ -0,0 +1,342 @@
...
+// ▗▄ ▄▖                         
+// ▐█ █▌      ▐▌                 
+// ▐███▌ ▟█▙ ▐███  ▟█▙  █▟█▌▗▟██▖
+// ▐▌█▐▌▐▛ ▜▌ ▐▌  ▐▛ ▜▌ █▘  ▐▙▄▖▘
+// ▐▌▀▐▌▐▌ ▐▌ ▐▌  ▐▌ ▐▌ █    ▀▀█▖
+// ▐▌ ▐▌▝█▄█▘ ▐▙▄ ▝█▄█▘ █   ▐▄▄▟▌
+// ▝▘ ▝▘ ▝▀▘   ▀▀  ▝▀▘  ▀    ▀▀▀
+// ----------------------------------------------------------------
+
+// Motor globals
```
Comment (anthonyjclark) [inline _arduino/nimbleProportionalControl.ino]:
```suggestion
```

File: _arduino/nimbleProportionalControl.ino [hunk 23]
```diff
@@ -0,0 +1,342 @@
...
+#define PULSES_PER_ROTATION 360
+#define SPEED_THRESH_PERCENT 5  // in terms of absolute value of u, which goes from 0 to 100
+
+#define MOTOR_RIGHT_PWM D8  // yellow
+#define MOTOR_RIGHT_DIR D7  // white
+
+#define MOTOR_LEFT_PWM D10  // yellow
+#define MOTOR_LEFT_DIR D9   // white
+
+int MIN_PWM_VALUE = 0;
```
Comment (anthonyjclark) [inline _arduino/nimbleProportionalControl.ino]: Is this correct?

File: _arduino/nimbleProportionalControl.ino [hunk 24]
```diff
@@ -0,0 +1,342 @@
...
+int motorWebSpeedPercentLeft;
+
+int controlSignalLeft;
+int controlSignalRight;
+
+MotorController motorRightController(HIGH, MOTOR_RIGHT_DIR, MOTOR_RIGHT_PWM);
+MotorController motorLeftController(LOW, MOTOR_LEFT_DIR, MOTOR_LEFT_PWM);
+
+Chrono motorControlChrono;
+Chrono OLEDChrono;
```
Comment (anthonyjclark) [inline _arduino/nimbleProportionalControl.ino]: This doesn't below next to the motor code.

File: _arduino/nimbleProportionalControl.ino [hunk 25]
```diff
@@ -0,0 +1,342 @@
...
+
+// ----------------------------------------------------------------
```
Comment (anthonyjclark) [inline _arduino/nimbleProportionalControl.ino]: Move Bluetooth stuff into its own file.

File: _arduino/nimbleProportionalControl.ino [hunk 26]
```diff
@@ -0,0 +1,342 @@
...
+
+    if(OLEDChrono.hasPassed(OLED_UPDATE_PERIOD_MS)) {
+    OLEDChrono.restart();
+    u8x8.clear();
+    u8x8.print("left encoder speed: ");
+    u8x8.print(measuredSpeedLeft);
+    u8x8.print("\n");
+    u8x8.print("right encoder speed: ");
+    u8x8.print(measuredSpeedRight);
+    u8x8.print("\n");
```
Comment (anthonyjclark) [inline _arduino/nimbleProportionalControl.ino]: Move this outside of the motorControlChrono if statement.

File: _arduino/nimbleProportionalControl.ino [hunk 27]
```diff
@@ -0,0 +1,342 @@
...
+    motorRightController.set(directionValueRight, motorRightPWMValue);
+
+  }
+
+  
+  if (!deviceConnected && oldDeviceConnected) {  // device was connected but no longer true
+    Serial.println("Device disconnected.");
+    // turns both motors off immediately if it is disconnected (REVERSE is a placeholder, as the motor is set to 0 anyway)
+    motorRightController.set(REVERSE, 0);
+    motorLeftController.set(REVERSE, 0);
```
Comment (anthonyjclark) [inline _arduino/nimbleProportionalControl.ino]: Add a `stop` method to the motor controller and call that.

**Review by anthonyjclark — CHANGES_REQUESTED** [review]
Some requests:

* Please put this set of sketches into their own sub-folder.
* Also, the `motorControl.ino` file should use a `.h` extension.
* Format the code with the built-in code formatter

**Review by anthonyjclark — APPROVED** [review]

**Pull request merged**